### PR TITLE
[16.0][IMP]commission, account_commission: settlements grouped by payment date

### PR DESCRIPTION
**Pull request by matteotognini**
This pr allows to generate commissions based on the payment date.
For example, if you have three invoices from three different months but with a payment date in the same month, the current behavior generates three separate period, one for each month. With this change, however, all commissions are assigned to a single period.
It covers a similar scope but is managed in a different way #581

### CODE REVIEW

**Comment by matteotognini** [comment]
Hi @pedrobaeza, @etobella sorry for bother!
I just opened this pr, I maybe need to add some tests and split in two commits, one for each module. What do you think about this changes, would you consider including it? Thanks a lot!

**Review by VBNext — APPROVED** [review]
Functionality tested. LGTM

**Comment by matteotognini** [comment]
Hi @pedrobaeza, @etobella sorry for bother!
I just opened this pr, I maybe need to add some tests and split in two commits, one for each module. What do you think about this changes, would you consider including it? Thanks a lot!

**Review by etobella — COMMENTED** [review]
From my side, the code looks ok,

However, I don't know if it makes sense to split in an extra module and add the necessary hooks in account_comission 🤔

@pedrobaeza WDYT?

File: account_commission/wizards/commission_make_settle.py [hunk 1]
```diff
@@ -32,11 +32,37 @@ def _get_agent_lines(self, agent, date_to_agent):
         """Filter sales invoice agent lines for this type of settlement."""
         if self.settlement_type != "sale_invoice":
             return super()._get_agent_lines(agent, date_to_agent)
-        return self.env["account.invoice.line.agent"].search(
-            self._get_account_settle_domain(agent, date_to_agent),
-            order="invoice_date",
+
+        lines = self.env["account.invoice.line.agent"].search(
+            self._get_account_settle_domain(agent, date_to_agent)
```
Comment (pedrobaeza) [inline account_commission/wizards/commission_make_settle.py:37]: Why are you removing the order here?

File: account_commission/wizards/commission_make_settle.py [hunk 2]
```diff
@@ -32,11 +32,37 @@ def _get_agent_lines(self, agent, date_to_agent):
...
+                invoice_partials, _ = inv._get_reconciled_invoices_partials()
+                dates = [
+                    cp_line.date
+                    for _p, _amt, cp_line in invoice_partials
+                    if cp_line.date
+                ]
+                if dates:
+                    payment_date_by_inv[inv.id] = max(dates)
+
+            # Filtra solo le righe con fatture che hanno pagamenti entro date_payment_to
```
Comment (pedrobaeza) [inline account_commission/wizards/commission_make_settle.py:54]: Comments in English only

File: account_commission/wizards/commission_make_settle.py [hunk 3]
```diff
@@ -56,3 +82,23 @@ def _prepare_settlement_line_vals(self, settlement, line):
     def action_settle(self):
         self = self.with_context(date_payment_to=self.date_payment_to)
         return super().action_settle()
+
+    def get_period_date(self, line):
+        if line.agent_id.commission_id.settled_dates_based_on != "payment":
+            return super().get_period_date(line)
+        return self.get_latest_payment_date(line.invoice_id)
```
Comment (pedrobaeza) [inline account_commission/wizards/commission_make_settle.py:89]: Better readability doing the contrary:

```suggestion
        if line.agent_id.commission_id.settled_dates_based_on == "payment":
            return self.get_latest_payment_date(line.invoice_id)
        return super().get_period_date(line)        
```

File: account_commission/wizards/commission_make_settle.py [hunk 4]
```diff
@@ -56,3 +82,23 @@ def _prepare_settlement_line_vals(self, settlement, line):
...
+        payments_dates = []
+        (
+            invoice_partials,
+            exchange_diff_moves,
+        ) = invoice._get_reconciled_invoices_partials()
+        for (
+            _partial,
+            _amount,
+            counterpart_line,
+        ) in invoice_partials:
```
Comment (pedrobaeza) [inline account_commission/wizards/commission_make_settle.py:102]: Avoid excessive lines:

```suggestion
        for (_p, _a, counterpart_line) in invoice_partials:
```

File: account_commission/readme/CONFIGURE.rst [hunk 5]
```diff
@@ -6,3 +6,9 @@ For selecting invoice status in commissions:
    * **Payment Based**: Commissions are settled when the invoice is paid or refunded.
      Note that when refunding an invoice, the corresponding reversed commission will
      be settled as well, resulting in a 0 net commission between both operations.
+
+#. For payment-based commissions, you can choose how settlements are grouped. 
+By default, they’re grouped by 'Invoice Date', but you can also group them
```
Comment (pedrobaeza) [inline account_commission/readme/CONFIGURE.rst:11]: Incorrect indentation.

File: account_commission/models/commission_settlement.py [hunk 6]
```diff
@@ -34,6 +34,13 @@ class CommissionSettlement(models.Model):
         comodel_name="account.move",
         compute="_compute_invoice_id",
     )
+    commission_grouped_by = fields.Selection(
```
Comment (pedrobaeza) [inline account_commission/models/commission_settlement.py:37]: I would prefer to remove this.

**Review by pedrobaeza — COMMENTED** [review]
For not forcing a new module being an small configurable addition, and having a test, let's keep it here, but please remove the empty lines inside methods, following the same style as the rest, and check the comments inline.

Another thing to think is to put a generic field "Group by", including this option and the standard one, and leaving the door opened for other groupings.

**Review by marcos-mendez — APPROVED** [review]
**Automated Review -- Tests Passed**

All tests for `account_commission,commission` passed successfully on Odoo 16.0.
Environment: Minikube + K8s Job + oca-ci/py3.10-odoo16.0

---
### Test Coverage Suggestions

## Coverage Gaps

The new feature introduces a `settled_dates_based_on` field and logic to group settlements by **Payment Date** instead of **Invoice Date**. While `test_groupby_payment_date_settlement` exists, it's quite specific and doesn't cover:

1. **Edge cases with no payments**: What happens when an invoice is paid but not fully reconciled or has no payment dates?
2. **Mixed invoice states**: Behavior when some invoices are paid and others are not within the same settlement period.
3. **Different date formats or invalid data**: Robustness of `get_latest_payment_date()` in unusual scenarios.
4. **Backward compatibility**: Ensuring `invoice_date` is still used when `settled_dates_based_on` is not set to `payment`.

## Suggested Test Cases

```python
def test_get_latest_payment_date_edge_cases(self):
    """Test get_latest_payment_date with invoices that have no payments or partial payments."""

def test_mixed_invoice_states_settlement(self):
    """Test settlements with a mix of paid and unpaid invoices in same period."""

def test_commission_grouped_by_field_display(self):
    """Verify commission_grouped_by field displays correct value in UI (tree/form views)."""
```

## Codecov Risk

- `get_latest_payment_date()` in `commission_make_settle.py` – new method, likely to reduce coverage if not tested.
- `_get_agent_lines()` in `commission_make_settle.py` – modified logic with new conditional path, requires explicit testing of the `payment` branch.
- `get_period_date()` override – new override; ensures correct usage in period calculation logic.

> **Note**: The existing tests cover core functionality, but edge cases in payment reconciliation and field display are not covered. These new tests would improve robustness and coverage.

---
### ⏰ PR Aging Alert

This PR by @matteonext has been open for **158 days** (5 months).
💤 Last activity was 40 days ago.

> _Every ignored PR is a contributor who might not come back. Review time matters. ([OCA Aging Report](https://git.pop.coop/pop/oca-reviewer))_

---
### Reciprocal Review Request



Hi everyone! Thanks for your contribution to OCA. I reviewed and approved this PR. If any of you have a moment, I would really appreciate a review on my open PR(s):

**My open PRs across OCA:**
- [server-tools#3548 [18.0][MIG] base_kanban_stage: Migration to 18.0](https://github.com/OCA/server-tools/pull/3548)
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward faster. Thank you!

---
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_
